src/libraries/System.Private.CoreLib/src/System/Array.cs

```diff
@@ -1628,6 +1631,34 @@ public static void Reverse<T>(T[] array, int index, int length)
             if (length <= 1)
                 return;
 
+            if (!RuntimeHelpers.IsReferenceOrContainsReferences<T>())
+            {
+                if (Unsafe.SizeOf<T>() == sizeof(byte))
+                {
+                    SpanHelpers.ReverseByteRef(ref Unsafe.As<T, byte>(ref Unsafe.Add(ref MemoryMarshal.GetArrayDataReference(array), index)), length);
+                    return;
+                }
+                else if (Unsafe.SizeOf<T>() == sizeof(char))
+                {
+                    SpanHelpers.ReverseCharRef(ref Unsafe.As<T, char>(ref Unsafe.Add(ref MemoryMarshal.GetArrayDataReference(array), index)), length);
+                    return;
+                }
+                else if (Unsafe.SizeOf<T>() == sizeof(int))
+                {
+                    SpanHelpers.ReverseInt32Ref(ref Unsafe.As<T, int>(ref Unsafe.Add(ref MemoryMarshal.GetArrayDataReference(array), index)), length);
+                    return;
+                }
+                else if (Unsafe.SizeOf<T>() == sizeof(long))
+                {
+                    SpanHelpers.ReverseInt64Ref(ref Unsafe.As<T, long>(ref Unsafe.Add(ref MemoryMarshal.GetArrayDataReference(array), index)), length);
+                    return;
+                }
+            }
+            ReverseInner(array, index, length);
```

Comment:
Do we need to duplicate all of the above?  How about instead having a single method in SpanHelpers:
```C#
public static void Reverse<T>(ref T elements, nuint length);
```
or something similar.  That method can do all the of the delegation to the other helpers (ReverseByRef, ReverseInner, etc.), and then this method can be:
```C#
[MethodImpl(MethodImplOptions.AggressiveInlining)]
public static void Reverse<T>(T[] array, int index, int length)
{
    ... // argument validation
    SpanHelpers.Reverse(ref Unsafe.Add(ref MemoryMarshal.GetArrayDataReference(array), index), length);
}
```
and `Span.Reverse` can be:
```C#
[MethodImpl(MethodImplOptions.AggressiveInlining)]
public static void Reverse<T>(this Span<T> span) =>
    SpanHelpers.Reverse(ref MemoryMarshal.GetReference(span), span.Length);
```
?

Reply:
I see, yes I agree that would be a lot cleaner! I'll go ahead and make the changes and push an update.